### PR TITLE
Disable NVML monitoring on WSL

### LINKED ISSUE #5567: `has_cuda_context` fails in WSL
**What happened**:
When attempting to check if a CUDA context has already been created using `has_cuda_context` in WSL, we see failures at the NVML level:

```python
NVMLError_Unknown                         Traceback (most recent call last)
<ipython-input-2-086bc6eeaba8> in <module>
----> 1 has_cuda_context()

~/dev/rapids-dask/distributed/distributed/diagnostics/nvml.py in has_cuda_context()
     76         handle = pynvml.nvmlDeviceGetHandleByIndex(index)
     77         if hasattr(pynvml, "nvmlDeviceGetComputeRunningProcesses_v2"):
---> 78             running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses_v2(handle)
     79         else:
     80             running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)

~/miniconda3/envs/rapids-dask/lib/python3.8/site-packages/pynvml/nvml.py in nvmlDeviceGetComputeRunningProcesses_v2(handle)
   2139     else:
   2140         # error case
-> 2141         raise NVMLError(ret)
   2142 
   2143 def nvmlDeviceGetComputeRunningProcesses(handle):
```

This is actually expected, as [querying active processes is not yet supported in WSL](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#features-not-yet-supported).

**What you expected to happen**:
We would want the function to work as desired, i.e. returning false or the index of the device for which there's a CUDA context.

**Minimal Complete Verifiable Example**:

```python
from distributed.diagnostics.nvml import has_cuda_context

has_cuda_context()
```

The error also occurs running the bare PyNVML code:

```python
from pynvml import *

nvmlInit()

h = nvmlDeviceGetHandleByIndex(0)
nvmlDeviceGetComputeRunningProcesses_v2(h)
```

**Anything else we need to know?**:
For context, this causes issues when attempting to create a CUDA context in dask-cuda, which is currently being tracked in https://github.com/rapidsai/dask-cuda/issues/816.

I think a potential solution to this issue, in the same vein as https://github.com/dask/distributed/pull/5343, would be to make a special case for WSL in `has_cuda_context`. This could either be done implicitly with a try/except block to catch the NVML error, or explicitly by checking if the OS is WSL and doing something different in that case:

```python
# https://www.scivision.dev/python-detect-wsl/
from platform import uname

def in_wsl() -> bool:
    return 'microsoft-standard' in uname().release
```

**Environment**:

- Dask version: latest `main`
- Python version: 3.8
- Operating System: ubuntu 20.04 (WSL2)
- Install method (conda, pip, source): source

cc @pentschev

### PR DESCRIPTION
**Pull request by charlesbluca**
Adds a check `_in_wsl()` to check if Distributed is being run from a WSL Linux subshell, and if so disables NVML monitoring. This is due to the fact that WSL's NVML does not currently support all queries, which can result in a variety of issues when starting up a cluster for both Distributed and Dask-CUDA; see https://docs.nvidia.com/cuda/wsl-user-guide/index.html#features-not-yet-supported.

Note that with this PR, we should be able to revert https://github.com/dask/distributed/pull/5343, as we no longer have to worry about real-time NVML metric tracking on WSL. However, I think it would probably be better to leave those checks in for the case where all NVML queries in WSL work _except_ for GPU utilization.

cc @pentschev 

- [x] Closes #5567
- [x] Closes https://github.com/rapidsai/dask-cuda/issues/816
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/diagnostics/nvml.py
Comment (charlesbluca) [inline distributed/diagnostics/nvml.py:94]: Once we get to this point, should we do something to either disable NVML monitoring or inform the user of its need to be disabled?

Comment (pentschev) [reply distributed/diagnostics/nvml.py:94]: I don't think the issues are necessarily related, it seems to me that the issue with monitoring is a bug, whereas this one is a known missing feature. Therefore, I don't think it makes sense to do either of those things at this place exactly. However, we could add some warning if we identify WSL is being used so that the user is aware of the bug and let know that it can be disabled via the environment variable, or alternatively we could disable NVML altogether when the system is using WSL and in this case this PR would possibly be unnecessary.

**Review by pentschev — APPROVED** [review]
LGTM, thanks @charlesbluca !

File: distributed/diagnostics/nvml.py
Comment (charlesbluca) [reply distributed/diagnostics/nvml.py:94]: Hmm that actually does seem like a more elegant solution here - just adding `_in_wsl()` to this check:

https://github.com/dask/distributed/blob/d0b40d3b9c36e4ebcd61f95baab94c6344e17f1d/distributed/diagnostics/nvml.py#L18-L20

And adding the same general check for NVML initialization that we use for `device_get_count`:

https://github.com/dask/distributed/blob/d0b40d3b9c36e4ebcd61f95baab94c6344e17f1d/distributed/diagnostics/nvml.py#L38-L40

Is sufficient to resolve all WSL issues, including those noted in https://github.com/dask/distributed/issues/5342 since we no longer need to worry about monitoring NVML metrics for non-GPU clusters. The only downside I could imagine with this approach is if there were users who wanted to view the GPU metrics of workers when using a `LocalCluster`, though those metrics wouldn't really be indicative of anything in the computations.

Comment (pentschev) [reply distributed/diagnostics/nvml.py]: But this is disabling WSL completely, until we can verify all the metrics are available and eventually undo changes here, right? Meaning, if some metric was usable before, it can't be visualized at all anymore.

Besides what you already pointed out, another downside I see with this is if we eventually decide to reenable NVML diagnostics in WSL, we may still end up needing to add the check for `has_cuda_context` again, as we don't know whether all limitations will be resolved at the same time. I don't have strong opinions either way, just trying to point out potential effects we may experience in the future.

Comment (charlesbluca) [reply distributed/diagnostics/nvml.py]: > Meaning, if some metric was usable before, it can't be visualized at all anymore.

Correct; I originally assumed that this wouldn't be much of an issue considering that `LocalCUDAClusters` can't monitor these metrics anyways on WSL, and the metrics aren't really relevant for a standard `LocalCluster` (pretty sure all the workers just monitor the first sequential GPU?), but I haven't checked to see if it's possible to do NVML monitoring on other cluster types using `CUDAWorkers`.

> we may still end up needing to add the check for has_cuda_context again, as we don't know whether all limitations will be resolved at the same time.

That's a good point - if it makes sense to, I can add back the try/except for `has_cuda_context`, maybe with a note that the except block shouldn't be getting reached at this point, but might be relevant later.

Comment (pentschev) [reply distributed/diagnostics/nvml.py]: > Correct; I originally assumed that this wouldn't be much of an issue considering that LocalCUDAClusters can't monitor these metrics anyways on WSL, and the metrics aren't really relevant for a standard LocalCluster (pretty sure all the workers just monitor the first sequential GPU?), but I haven't checked to see if it's possible to do NVML monitoring on other cluster types using CUDAWorkers.

Why can't `LocalCUDACluster`s monitor those metrics on WSL? Maybe I'm missing something, but I thought it was just a few of them and only due to the known bug.

> That's a good point - if it makes sense to, I can add back the try/except for has_cuda_context, maybe with a note that the except block shouldn't be getting reached right now, but might be relevant later.

I don't know if that's necessary, it may be useful to prevent us from potentially debugging the same issue again in the future? I'm not certain what would be the correct course of action here, so I'll leave up to you to decide about this one.

Comment (charlesbluca) [reply distributed/diagnostics/nvml.py]: > Why can't `LocalCUDAClusters` monitor those metrics on WSL?

The error in https://github.com/dask/distributed/issues/5567#issuecomment-988096144 seemed to imply that after unblocking the issue with checking for CUDA context, we are unable to get device handles by index, and would need to set `DASK_DISTRIBUTED__DIAGNOSTICS__NVML=False` in order to continue using the CUDA cluster without getting constant errors from the attempts to monitor NVML metrics.

> I'm not certain what would be the correct course of action here, so I'll leave up to you to decide about this one.

I think leaving out the try/except block, but including a comment referencing the original issue should be sufficient.

Comment (pentschev) [reply distributed/diagnostics/nvml.py]: > The error in #5567 (comment) seemed to imply that after unblocking the issue with checking for CUDA context, we are unable to get device handles by index, and would need to set DASK_DISTRIBUTED__DIAGNOSTICS__NVML=False in order to continue using the CUDA cluster without getting constant errors from the attempts to monitor NVML metrics.

Is that _always_ the case? I thought NVML diagnostics in WSL worked in some cases, am I mistaken? If I am, then I agree it's sufficient.

> I think leaving out the try/except block, but including a comment referencing the original issue should be sufficient.

Agreed, a comment would be good and sufficient for now.

Comment (charlesbluca) [reply distributed/diagnostics/nvml.py]: > Is that _always_ the case?

Not always; to sum up the cases I know of where you can view real-time NVML metrics on WSL:

- using a `LocalCluster` without this PR; all the workers will be monitoring the first sequential GPU, which shouldn't be impacted since computation is happening on CPU
- using a `LocalCUDACluster` without this PR; immediately after construction, the cluster seems to monitor the relevant GPUs without issue. If you try to do anything with CUDA afterwards (i.e. construct a cuDF / dask-cuDF dataframe), the NVML monitoring will being failing

Given those cases, I feel like the choice to disable monitoring is reasonable in that:

- for a `LocalCluster`, these metrics generally don't have any relevance to the computations on CPU
- for a `LocalCUDACluster`, we generally can't monitor metrics once we do any computations on GPU, which in a sense defeats the purpose of having the metrics in the first place

That being said, I haven't tried any other set ups; I hypothesize that the only way we would be able to have working GPU workers without running into this issue is if we created them in a way that avoids a call to `has_cuda_context`, though I'd need to take some to think about if this is possible.

Comment (pentschev) [reply distributed/diagnostics/nvml.py]: Thanks for all details, Charles. Given all the issues you pointed out, I agree it's reasonable to just disable it entirely instead of trying to patch around every single minor issue for something that ultimately has to be fixed by NVML.

With that said, I think we're good here. Thanks for the patience to summarize everything!

**Review by pentschev — APPROVED** [review]
Latest changes look good to me, thanks @charlesbluca . Would you mind taking a look/approving as well @quasiben ?

**Comment by quasiben** [comment]
This looks good to me.  Thanks @pentschev for reviewing and @charlesbluca for doing the work.   I want to check with @jrbourbeau before merging in as I thought we were going to do a release this week.  To be clear this shouldn't impact any release issues as the code changes are quite limited to WSL and GPUs -- still, we should get confirmation before merging in

**Review by jrbourbeau — COMMENTED** [review]
@quasiben is correct, there's a release scheduled for this Friday (xref https://github.com/dask/community/issues/209). FWIW I suspect this release will be the last one for the year (barring any urgent issues arising). The changes here look reasonable to me, but I know handling NVML initialization can be tricky. Do folks want to include this in the upcoming release? Or is it okay to wait until Monday to merge this?

Also, does anyone have access to a machine running WSL that can test the changes here? (again, they looks reasonable, but it would be good to test before merging if posible)

**Comment by charlesbluca** [comment]
> Do folks want to include this in the upcoming release?

I'm okay with pushing this back to next release if that's the safe option - I haven't seen too many issues on either Distributed or Dask-CUDA dealing with WSL NVML monitoring. so think it's safe to say this is a niche fix that isn't urgently needed.

> does anyone have access to a machine running WSL that can test the changes here?

I made and tested these changes from a machine running CUDA-enabled WSL, primarily to unblock some Dask-CUDA related development of my own. I can verify that this unblocks some of the simpler cluster configs (i.e. vanilla `LocalCluster`, `LocalCUDACluster`, starting CUDA cluster using `dask-scheduler`/`dask-cuda-worker`), but haven't gone in depth to try out more niche configurations (i.e. `SSHCluster` using `CUDAWorkers`, `LocalCUDACluster` with UCX comms, etc.).

File: distributed/diagnostics/nvml.py
Comment (jakirkham) [inline distributed/diagnostics/nvml.py:20]: What happens if this runs on non-WSL Windows (not that we support that just curious)?

Comment (charlesbluca) [reply distributed/diagnostics/nvml.py:20]: Looks like on standard Windows this just outputs the version of the Windows kernel; for example, on Windows 11:

```python
>>> import platform
>>> platform.uname().release
'10'
```

**Comment by jrbourbeau** [comment]
rerun tests

**Review by jrbourbeau — APPROVED** [review]
Thanks all!

**Comment by lmeyerov** [comment]
FYI, we expect this general path to start getting used more, as wsl2 nvidia support means we expect devs to start switching from linux desktops / linux dual boots to windows-with-wsl2

**Pull request merged**
